Migrate several tests in integration-cli to integration

yongtang opened the pull request:
This fix moves TestAPIOptionsRoute and TestAPIGetEnabledCORS to new integration tests

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Migrate several tests in integration-cli to integration

This fix moves TestAPIOptionsRoute and TestAPIGetEnabledCORS to new integration tests

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

Remove unneeded comments

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

AkihiroSuda commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,29 @@
+	// and here none of the headers below are in the response?
+	//c.Log(res.Header)
+	//c.Assert(res.Header.Get("Access-Control-Allow-Origin"), check.Equals, "*")
+	//c.Assert(res.Header.Get("Access-Control-Allow-Headers"), check.Equals, "Origin, X-Requested-With, Content-Type, Accept, X-Registry-Auth")
```
LGTM, but probably these old comments can be removed?

yongtang replied:
@AkihiroSuda Updated, thanks.

thaJeztah commented on integration-cli/docker_api_test.go:
```diff
@@ -15,24 +15,6 @@ import (
-	// TODO: @runcom incomplete tests, why old integration tests had this headers
-	// and here none of the headers below are in the response?
-	//c.Log(res.Header)
-	//assert.Equal(c, res.Header.Get("Access-Control-Allow-Origin"), "*")
```
Looks like this behaviour changed in https://github.com/moby/moby/pull/32174, where the middleware was only added if a non-empty value is set for the CORS headers.

Interestingly though, the middleware has a special case for empty strings, and changes that to use `"*"`; https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/api/server/middleware/cors.go#L24-L35

So I'm wondering if 

- the `if cfg.CorsHeaders != ""` should be removed 🤔 (OTOH, `*` and "not set" could be considered equivalent?)
- otherwise; if the special case (change `""` to `"*"`) should be removed

thaJeztah commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,24 @@
+}
+
+func TestAPIGetEnabledCORS(t *testing.T) {
+	resp, body, err := req.Get("/version")
```
This could be changed to `/_ping` perhaps (as we're not interested in the `version` output, just in the headers for any response)

thaJeztah commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,24 @@
+	assert.Equal(t, resp.StatusCode, http.StatusOK)
+}
+
+func TestAPIGetEnabledCORS(t *testing.T) {
```
Looks like this test is currently doing anything, and it was originally written to test that the CORS headers were added (which defaulted to `*`); also see my comment above.

This is the commit that added the original version of the test (it moved around a few times); https://github.com/moby/moby/commit/6d5bdff3942ce5e030b2cbd1510f418de25a1a53, which was part of https://github.com/moby/moby/pull/806


Pending the options in my earlier comment, the `if` should be removed, or otherwise this test should start a new daemon with the correct flag set.

Removing the `if`, or starting the daemon with `--api-cors-header='*'` both give;

```bash
curl --unix-socket /var/run/docker.sock http://local/_ping -v
*   Trying /var/run/docker.sock...
* Connected to local (/var/run/docker.sock) port 80 (#0)
> GET /_ping HTTP/1.1
> Host: local
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, X-Registry-Auth
< Access-Control-Allow-Methods: HEAD, GET, POST, DELETE, PUT, OPTIONS
< Access-Control-Allow-Origin: *
< Api-Version: 1.41
< Cache-Control: no-cache, no-store, must-revalidate
< Docker-Experimental: false
< Ostype: linux
< Pragma: no-cache
< Server: Docker/dev (linux)
< Date: Mon, 19 Aug 2019 09:51:33 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
< 
* Curl_http_done: called premature == 0
* Connection #0 to host local left intact
```

thaJeztah reviewed (changes requested):
left some comments; looks like one of the tests is currently broken (or the daemon behavior is incorrect)

yongtang committed:
Update TestAPIGetEnabledCORS to set the deamon flag,

This fix updates the TestAPIGetEnabledCORS test, to
create a new daemon with the right flag.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@thaJeztah The PR has been updated. Please take a look.

thaJeztah commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,51 @@
+	"time"
+
+	"github.com/docker/docker/internal/test/daemon"
+
```
nit: remove this line and put the above import in the same group

thaJeztah commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,51 @@
+	assert.Equal(t, resp.StatusCode, http.StatusOK)
+
+	assert.Equal(t, len(resp.Header["Access-Control-Allow-Origin"]), 1)
+	assert.Equal(t, resp.Header["Access-Control-Allow-Origin"][0], "'*'")
```
The value we're checking for here is incorrect; the header should not include the single quotes;

```suggestion
	assert.Equal(t, resp.Header["Access-Control-Allow-Origin"][0], "*")
```

curious why you didn't pick the old approach here;

```suggestion
	assert.Equal(t, resp.Header.Get("Access-Control-Allow-Origin"), "'*'")
```

thaJeztah commented on integration/system/options_test.go:
```diff
@@ -0,0 +1,51 @@
+
+	d := daemon.New(t)
+
+	d.Start(t, "--api-cors-header='*'")
```
The problem here is that this works when executed in a shell (where the shell would strip the single quotes (`'`)  (but also expand `*` to files in the same directory), but the integration test is not executed in a shell; because of that, the quotes are not stripped.

However, because there's no shell, there's also no need to use quotes (because `*` are not expanded / glob'ed), so this will work:

```suggestion
	d.Start(t, "--api-cors-header=*")
```

Or, this:

```suggestion
	d.Start(t, "--api-cors-header, "*")
```

Double checked; with that change (and the updated `assert`); the test still works;

```
make TEST_FILTER=TestAPIGetEnabledCORS test-integration

....


=== RUN   TestAPIGetEnabledCORS
--- PASS: TestAPIGetEnabledCORS (0.75s)
    options_test.go:28: Creating a new daemon at: /go/src/github.com/docker/docker/bundles/test-integration/TestAPIGetEnabledCORS
    daemon.go:336: [ddd936e8e062d] waiting for daemon to start
    daemon.go:336: [ddd936e8e062d] waiting for daemon to start
    daemon.go:364: [ddd936e8e062d] daemon started
    daemon.go:472: [ddd936e8e062d] Stopping daemon
    daemon.go:307: [ddd936e8e062d] exiting daemon
    daemon.go:459: [ddd936e8e062d] Daemon stopped
PASS
```

thaJeztah reviewed (changes requested):
Thanks for updating; found some issues (commented below)
